### ISSUE TITLE
Metal grammar decode: completed grammar state misclassified as GrammarConstraintBlocked (strict JSON 500 on success)

### ISSUE THREAD

**Issue by ohdearquant** [ohdearquant/lattice#1064]
### Symptom

A strict `json_schema` generation on the Metal path that produces a complete, valid document can be rejected as `GrammarConstraintBlocked`, which `lattice_serve` maps to `blocked_constraint` / HTTP 500. The generation actually succeeded; the error fires one step later.

### Root cause: the CPU/Metal grammar contract is only half-mirrored

The CPU paths implement the #611 fail-closed contract in two halves:

1. **Fail closed** when the grammar blocks every token in a state that is *not* complete (`has_finite_logit` guard).
2. **Terminate cleanly** when the state is complete with no legal continuation — both by checking `engine.is_complete_without_continuation(gs)` after every `advance` (e.g. `crates/inference/src/model/qwen35/generation.rs:222`) and by exempting accepting states at every all-blocked guard (`generation.rs:189`, 640, 815, 1023, 1910, 2038):

```rust
if !has_finite_logit(&scratch.logits[..cfg.vocab_size]) {
    if engine.is_complete_without_continuation(gs) {
        return Ok(grammar_output(...));   // clean stop
    }
    return Err(InferenceError::GrammarConstraintBlocked(...));
}
```

The Metal path (`crates/inference/src/forward/metal_qwen35.rs`) mirrors only half 1. All six `GrammarConstraintBlocked` sites (9247, 9408, 14183, 14354, 17081, 17292) fail closed unconditionally, and the file contains no call to `is_complete_without_continuation` (or any completion check) at all.

Sequence on Metal for a schema whose accepting state has no legal continuation (typical strict JSON: nothing may follow the final `}`):

1. Model samples the final `}`; `advance` succeeds; loop continues.
2. Next iteration: `forward_step` + `mask_logits` on the now-complete state masks **every** token to `-inf` (the engine has no EOS special-casing; EOS simulates as illegal bytes and empty-byte tokens are always blocked).
3. `!has_finite_logit(...)` → `Err(GrammarConstraintBlocked)` → `blocked_constraint` / 500.

There is also no post-`advance` completion break in the Metal decode loops, so the completed state always survives to the next mask. Note the mask itself never leaves EOS legal under an active grammar, so a grammar-constrained Metal generation cannot terminate via `GrammarStop`/`Eos` either; completion detection is the only clean exit, and it is missing.

### Why tests did not catch it

The Metal grammar tests assert the fail-closed half (all-blocked at step 0 and in the decode loop → typed error, e.g. the mutation-sensitivity tests around `metal_qwen35.rs:34808`), which is correct for *incomplete* states. No test drives a grammar to completion on the Metal path and asserts a successful return.

### Fix shape

Mirror the CPU contract at the Metal sites:

- At each decode-loop all-blocked guard (9408, 14354, 17292): completed state → clean stop with the accumulated text (stop reason `Grammar`), otherwise keep the typed error.
- Step-0 sites (9247, 14183, 17081): same exemption for schemas whose initial state is already accepting (empty-accepting grammars), matching `generation.rs:189`.
- Add the post-`advance` `is_complete_without_continuation` break to the Metal decode loops so a completed generation stops without paying one extra forward step, matching `generation.rs:222`.
- Regression test: fixture grammar that completes after N tokens; assert Metal returns success with the full text, plus an end-to-end strict `json_schema` 200 through the serve path. Test must fail if the completion exemption is reverted.

### Severity

High for structured output on the Metal path: a strict-schema request either completes (and can 500 as above) or truncates on budget. Affects the production decode path; discovered during an adversarial review round, verified at source on `main@e1000d310`.

**Comment by ohdearquant** [comment]
Fix landed in #1067, squash-merged to `integration/perf-regression-fixes` at e78cb9806c33370b3d512bb4c808453c80ecbc18. All three Metal decode paths now stop with `finish_reason: "stop"` (`StopReason::Grammar`) when the grammar state completes, matching the CPU contract; step-0 guards exempt an already-complete state. Verified with mutation-sensitive tests (the three rewritten success tests fail with the original `GrammarConstraintBlocked` when the completion checks are mutated out) plus the full Metal and CPU grammar sweeps. This issue closes when the integration branch merges to main. Follow-up still open: a serve-level e2e test asserting strict json_schema on the Metal backend returns HTTP 200.
